satpy/dataset/data_dict.py

```diff
@@ -180,25 +180,24 @@ def get(self, key, default=None):
         return super(DatasetDict, self).get(key, default)
 
     def __setitem__(self, key, value):
-        """Support assigning 'Dataset' objects or dictionaries of metadata."""
-        if hasattr(value, 'attrs'):
-            # xarray.DataArray objects
-            value_info = value.attrs
-        else:
-            value_info = value
+        """Support assigning 'Dataset' objects or dictionaries of metadata.
+
+        Args:
+            key (Union[str, DataID]): The key to assign the value (xarray.DataArray) to.
+            value (xarray.DataArray): The 'Dataset' to be assigned.
+        """
+        value_info = value.attrs
+
         # use value information to make a more complete DataID
         if not isinstance(key, DataID):
-            key = self._create_dataid_key(key, value_info)
+            key = DataID.from_dataarray(value, default_keys={k: v for k, v in default_id_keys_config.items() if k
+                                                             in value_info.keys()})
 
         # update the 'value' with the information contained in the key
-        try:
-            new_info = key.to_dict()
-        except AttributeError:
-            new_info = key
+        new_info = key.to_dict()
         if isinstance(value_info, dict):
```

Comment:
I think value_info should always be a dict since it's the `xr.DataArrays` attributes... unless the DataArray has no attributes. Mmh need to check that. Should probably then be covered by tests too...

Reply:
So either the user didn't provide a DataArray (numpy, dask, etc), in which case there would be no `.attrs` attribute (an exception above). Or we require a DataArray as input and we know it has `.attrs` and this `isinstance` check can be removed. Right?